v2/stream/datav2.go

```diff
@@ -0,0 +1,482 @@
+
+	// concurrency
+	readerOnce    sync.Once
+	wsWriteMutex  sync.Mutex
+	wsReadMutex   sync.Mutex
+	handlersMutex sync.RWMutex
+}
+
+func newDatav2Stream() *datav2stream {
+	if s := os.Getenv("DATA_PROXY_WS"); s != "" {
+		DataStreamURL = s
+	}
+	stream = &datav2stream{
+		feed:          "iex",
+		authenticated: atomic.Value{},
+		tradeHandlers: make(map[string]func(trade Trade)),
+		quoteHandlers: make(map[string]func(quote Quote)),
+		barHandlers:   make(map[string]func(bar Bar)),
+	}
+
+	stream.authenticated.Store(false)
+	stream.closed.Store(false)
+
+	return stream
+}
+
+func (s *datav2stream) useFeed(feed string) error {
+	feed = strings.ToLower(feed)
+	switch feed {
+	case "iex", "sip":
+	default:
+		return errors.New("unsupported feed: " + feed)
+	}
+	if s.feed == feed {
+		return nil
+	}
+	s.feed = feed
+	if s.conn == nil {
+		return nil
+	}
+	// we are already connected to the wrong feed
+	// to restart it we close the stream and readForever will do the reconnect
+	return s.close(false)
+}
+
+func (s *datav2stream) subscribeTrades(handler func(trade Trade), symbols ...string) error {
+	if err := s.ensureRunning(); err != nil {
+		return err
+	}
+
+	if err := s.sub(symbols, nil, nil); err != nil {
+		return err
+	}
+
+	s.handlersMutex.Lock()
+	defer s.handlersMutex.Unlock()
+
+	for _, symbol := range symbols {
+		s.tradeHandlers[symbol] = handler
+	}
+
+	return nil
+}
+
+func (s *datav2stream) subscribeQuotes(handler func(quote Quote), symbols ...string) error {
+	if err := s.ensureRunning(); err != nil {
+		return err
+	}
+
+	if err := s.sub(nil, symbols, nil); err != nil {
+		return err
+	}
+
+	s.handlersMutex.Lock()
+	defer s.handlersMutex.Unlock()
+
+	for _, symbol := range symbols {
+		s.quoteHandlers[symbol] = handler
+	}
+
+	return nil
+}
+
+func (s *datav2stream) subscribeBars(handler func(bar Bar), symbols ...string) error {
+	if err := s.ensureRunning(); err != nil {
+		return err
+	}
+
+	if err := s.sub(nil, nil, symbols); err != nil {
+		return err
+	}
+
+	s.handlersMutex.Lock()
+	defer s.handlersMutex.Unlock()
+
+	for _, symbol := range symbols {
+		s.barHandlers[symbol] = handler
+	}
+
+	return nil
+}
+
+func (s *datav2stream) unsubscribe(trades []string, quotes []string, bars []string) error {
+	if err := s.ensureRunning(); err != nil {
+		return err
+	}
+
+	s.handlersMutex.Lock()
+	defer s.handlersMutex.Unlock()
+
+	for _, trade := range trades {
+		delete(s.tradeHandlers, trade)
+	}
+	for _, quote := range quotes {
+		delete(s.quoteHandlers, quote)
+	}
+	for _, bar := range bars {
+		delete(s.barHandlers, bar)
+	}
+
+	if err := s.unsub(trades, quotes, bars); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (s *datav2stream) close(final bool) error {
+	if s.conn == nil {
+		return nil
+	}
+
+	s.wsWriteMutex.Lock()
+	defer s.wsWriteMutex.Unlock()
+
+	if err := s.conn.Close(websocket.StatusNormalClosure, ""); err != nil {
+		return err
+	}
+
+	if final {
+		s.closed.Store(true)
+	}
```

Comment:
Would it make sense to move this above `if err := s.conn.Close(websocket.StatusNormalClosure, ""); err != nil {`?

I wonder what happens when the user tries to close the stream and receives an error on L183. Is the current outcome what the user would expect?